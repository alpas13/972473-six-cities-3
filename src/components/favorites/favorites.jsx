import React from "react";
import PropTypes from "prop-types";
import Header from "../header/header.jsx";
import OffersList from "../offers-list/offers-list.jsx";
import {uniqueFilter, filterByValue} from "../../utils";
import {favoritesStyle} from "../../const";

const Favorites = React.memo(function Favorites(props) {
  const {offers, authInfo, getFavoritesPage, getLoginPage} = props;
  return (
    <div className="page">
      <Header
        authInfo={authInfo}
        onEmailClick={getFavoritesPage}
        onSignInClick={getLoginPage}
      />
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {uniqueFilter(offers, `city`).map((city) => {
                return (
                  <li key={city} className="favorites__locations-items">
                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link" href="#">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>
                    <div className="favorites__places">
                      <OffersList
                        offers={filterByValue(offers, `city`, city)}
                        styleSettings={favoritesStyle}
                        authInfo={authInfo}
                        getLoginPage={getLoginPage}
                      />
                    </div>
                  </li>
                );
              })}
            </ul>
          </section>
        </div>
      </main>
      <footer className="footer container">
        <a className="footer__logo-link" href="main.html">
          <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33"/>
        </a>
      </footer>
    </div>
  );
});

Favorites.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
  authInfo: PropTypes.object,
  getFavoritesPage: PropTypes.func.isRequired,
  getLoginPage: PropTypes.func.isRequired,
};

export default Favorites;
