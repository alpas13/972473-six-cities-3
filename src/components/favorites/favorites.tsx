import * as React from "react";
import OffersList from "../offers-list/offers-list";
import {uniqueFilter, filterByValue} from "../../utils";
import {favoritesStyle} from "../../const";
import {connect} from "react-redux";
import {getFavorites} from "../../reducer/main/selectors";
import {Offer} from "../../types";

interface Props {
    offers: Offer[];
}

const Favorites: React.FC<Props> = (props: Props) => {
  const {offers} = props;
  return (
    <React.Fragment>
      {!!offers.length && <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {uniqueFilter(offers, `city`).map((city) => {
                return (
                  <li key={city} className="favorites__locations-items">
                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link" href="#">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>
                    <div className="favorites__places">
                      <OffersList
                        offers={filterByValue(offers, `city`, city)}
                        styleSettings={favoritesStyle}
                      />
                    </div>
                  </li>
                );
              })}
            </ul>
          </section>
        </div>
      </main>}
      {!!offers.length || <main className="page__main page__main--favorites page__main--favorites-empty">
        <div className="page__favorites-container container">
          <section className="favorites favorites--empty">
            <h1 className="visually-hidden">Favorites (empty)</h1>
            <div className="favorites__status-wrapper">
              <b className="favorites__status">Nothing yet saved.</b>
              <p className="favorites__status-description">Save properties to
                narrow down search or plan yor future trips.</p>
            </div>
          </section>
        </div>
      </main>}
    </React.Fragment>
  );
};

const mapStateToProps = (state) => ({
  offers: getFavorites(state),
});

export {Favorites};
export default connect(mapStateToProps)(React.memo(Favorites));
