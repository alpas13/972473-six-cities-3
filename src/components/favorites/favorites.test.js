import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Favorites} from "./favorites.jsx";

const initialState = {offers: [
  {
    id: 1,
    city: `Hamburg`,
    cityCoords: [52.3909553943508, 4.85309666406198],
    cityZoom: 9,
    propertyImage: [`path 1`],
    title: `Title text 1`,
    mark: `Premium`,
    previewImage: `path 1`,
    price: 1,
    bookmark: true,
    propertyText: `Text`,
    rating: {
      star: (Math.floor(4) / 5) * 100,
      value: 4,
    },
    features: {
      entire: `type text 1`,
      bedrooms: 1,
      adults: 1,
    },
    insideList: [
      `text 1`,
    ],
    host: {
      avatarUrl: `path 1`,
      id: 3,
      name: `text 1`,
      isPro: true,
    },
    coords: [
      52.3909553943508,
      4.85309666406198
    ],
    coordsZoom: 8,
  }
]};

const mockStore = configureStore();

test(`Correctly render Favorites component`, () => {
  const store = mockStore(initialState);
  const tree = renderer.create(
      <Favorites
        store={store}
        offers={initialState.offers}
        getFavoritesPage={() => {}}
      />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
