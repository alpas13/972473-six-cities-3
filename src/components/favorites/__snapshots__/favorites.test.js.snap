// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly render Favorites component 1`] = `
<main
  className="page__main page__main--favorites"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites"
    >
      <h1
        className="favorites__title"
      >
        Saved listing
      </h1>
      <ul
        className="favorites__list"
      >
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Hamburg
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="favorites__card place-card"
              onMouseEnter={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="favorites__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="path 1"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      1
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button  button"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "80%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    Title text 1
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  type text 1
                </p>
              </div>
            </article>
          </div>
        </li>
      </ul>
    </section>
  </div>
</main>
`;
